independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanResolverImpl.java

```diff
@@ -81,7 +82,11 @@ public BeanInfo resolveAmbiguity(Set<BeanInfo> beans) {
 
     private List<BeanInfo> findMatching(TypeAndQualifiers typeAndQualifiers) {
         List<BeanInfo> resolved = new ArrayList<>();
-        for (BeanInfo b : beanDeployment.getBeans()) {
+        //optimisation for the simple class case
+        Collection<BeanInfo> potentialBeans = typeAndQualifiers.type.kind() == CLASS
```

Comment:
So in theory, even the parameterized types without wildcards and type variables should be safe, or?

Reply:
I was not sure, I went with the safest option.

Reply:
Ok, let's start with `CLASS`...